### ISSUE TITLE
error after update to 0.21.6

### ISSUE THREAD

**Issue by rsichny** [fastavro/fastavro#274]
I'm not sure what has changed in 0.21.6, but the following code totally worked in 0.21.5 and stopped working (message doesn't pass validation) after update:

```
import json
from collections import OrderedDict
from datetime import datetime

from fastavro import validate

schema = json.loads("""
{
    "type": "record",
    "name": "questionnaire",
    "namespace": "vela.messages.questionnaire",
    "fields": [
        {
            "type": {
                "type": "record",
                "major_version": 4,
                "minor_version": 10,
                "name": "message_header",
                "namespace": "vela.messages",
                "fields": [
                    {
                        "type": "string",
                        "name": "message_uuid"
                    },
                    {
                        "type": {
                            "type": "enum",
                            "name": "_message_source_name",
                            "namespace": "vela.messages",
                            "symbols": [
                                "questionnaire_api"
                            ]
                        },
                        "name": "message_source"
                    },
                    {
                        "type": {
                            "type": "array",
                            "items": "string"
                        },
                        "name": "producer_ip_addresses"
                    },
                    {
                        "type": {
                            "type": "long",
                            "logicalType": "timestamp-micros"
                        },
                        "name": "message_created_at"
                    },
                    {
                        "type": "long",
                        "name": "partner_internal_id"
                    },
                    {
                        "type": "long",
                        "name": "organization_internal_id"
                    },
                    {
                        "type": {
                            "type": "enum",
                            "name": "_business_event_type",
                            "namespace": "vela.messages",
                            "symbols": [
                                "QUESTIONNAIRE_CREATED",
                                "QUESTIONNAIRE_UPDATED",
                                "QUESTIONNAIRE_DELETED"
                            ]
                        },
                        "name": "business_event_type"
                    }
                ]
            },
            "name": "header"
        },
        {
            "type": {
                "type": "record",
                "name": "questionnaire",
                "namespace": "vela.model.questionnaire",
                "fields": [
                    {
                        "type": "long",
                        "name": "questionnaire_internal_id"
                    },
                    {
                        "type": "string",
                        "name": "label"
                    },
                    {
                        "type": {
                            "type": "map",
                            "values": "string"
                        },
                        "name": "display_label"
                    },
                    {
                        "type": [
                            "null",
                            "string"
                        ],
                        "name": "data_tag",
                        "default": null
                    },
                    {
                        "type": {
                            "type": "array",
                            "items": [
                                {
                                    "type": "record",
                                    "name": "_question",
                                    "namespace": "vela.model.questionnaire",
                                    "fields": [
                                        {
                                            "type": "long",
                                            "name": "questionnaire_node_internal_id"
                                        },
                                        {
                                            "type": {
                                                "type": "enum",
                                                "name": "_questionnaire_node_type_question",
                                                "namespace": "vela.model.questionnaire",
                                                "symbols": [
                                                    "QUESTION"
                                                ]
                                            },
                                            "name": "questionnaire_node_type"
                                        },
                                        {
                                            "type": "string",
                                            "name": "label"
                                        },
                                        {
                                            "type": {
                                                "type": "map",
                                                "values": "string"
                                            },
                                            "name": "text"
                                        },
                                        {
                                            "type": [
                                                "null",
                                                "string"
                                            ],
                                            "name": "data_tag",
                                            "default": null
                                        },
                                        {
                                            "type": [
                                                {
                                                    "type": "record",
                                                    "name": "_open_options",
                                                    "namespace": "vela.model.questionnaire",
                                                    "fields": [
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_question_type_open",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "OPEN"
                                                                ]
                                                            },
                                                            "name": "question_type"
                                                        },
                                                        {
                                                            "type": "boolean",
                                                            "name": "required",
                                                            "default": false
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                "long"
                                                            ],
                                                            "name": "min_length",
                                                            "default": null
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                "long"
                                                            ],
                                                            "name": "max_length",
                                                            "default": null
                                                        },
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_open_widget",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "INPUT",
                                                                    "TEXTAREA"
                                                                ]
                                                            },
                                                            "name": "widget"
                                                        }
                                                    ]
                                                },
                                                {
                                                    "type": "record",
                                                    "name": "_choice_options",
                                                    "namespace": "vela.model.questionnaire",
                                                    "fields": [
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_question_type_choice",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "CHOICE"
                                                                ]
                                                            },
                                                            "name": "question_type"
                                                        },
                                                        {
                                                            "type": "boolean",
                                                            "name": "required",
                                                            "default": false
                                                        },
                                                        {
                                                            "type": {
                                                                "type": "array",
                                                                "items": {
                                                                    "type": "record",
                                                                    "name": "choice_option",
                                                                    "namespace": "vela.model.types.questionnaire",
                                                                    "fields": [
                                                                        {
                                                                            "type": "string",
                                                                            "name": "value"
                                                                        },
                                                                        {
                                                                            "type": {
                                                                                "type": "map",
                                                                                "values": "string"
                                                                            },
                                                                            "name": "text"
                                                                        }
                                                                    ]
                                                                }
                                                            },
                                                            "name": "choices"
                                                        },
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_choice_widget",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "DROPDOWN",
                                                                    "RADIO",
                                                                    "BUTTON"
                                                                ]
                                                            },
                                                            "name": "widget"
                                                        },
                                                        {
                                                            "type": "long",
                                                            "name": "columns_number",
                                                            "default": 1
                                                        }
                                                    ]
                                                },
                                                {
                                                    "type": "record",
                                                    "name": "_multichoice_options",
                                                    "namespace": "vela.model.questionnaire",
                                                    "fields": [
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_question_type_multichoice",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "MULTICHOICE"
                                                                ]
                                                            },
                                                            "name": "question_type"
                                                        },
                                                        {
                                                            "type": "boolean",
                                                            "name": "required",
                                                            "default": false
                                                        },
                                                        {
                                                            "type": {
                                                                "type": "array",
                                                                "items": "vela.model.types.questionnaire.choice_option"
                                                            },
                                                            "name": "choices"
                                                        },
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_multichoice_widget",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "SELECT",
                                                                    "CHECKBOX",
                                                                    "BUTTON"
                                                                ]
                                                            },
                                                            "name": "widget"
                                                        },
                                                        {
                                                            "type": "long",
                                                            "name": "columns_number",
                                                            "default": 1
                                                        }
                                                    ]
                                                },
                                                {
                                                    "type": "record",
                                                    "name": "_datepicker_options",
                                                    "namespace": "vela.model.questionnaire",
                                                    "fields": [
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_question_type_datepicker",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "DATEPICKER"
                                                                ]
                                                            },
                                                            "name": "question_type"
                                                        },
                                                        {
                                                            "type": "boolean",
                                                            "name": "required",
                                                            "default": false
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                {
                                                                    "type": "int",
                                                                    "logicalType": "date"
                                                                }
                                                            ],
                                                            "name": "min_value",
                                                            "default": null
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                {
                                                                    "type": "int",
                                                                    "logicalType": "date"
                                                                }
                                                            ],
                                                            "name": "max_value",
                                                            "default": null
                                                        }
                                                    ]
                                                },
                                                {
                                                    "type": "record",
                                                    "name": "_datetimepicker_options",
                                                    "namespace": "vela.model.questionnaire",
                                                    "fields": [
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_question_type_datetimepicker",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "DATETIMEPICKER"
                                                                ]
                                                            },
                                                            "name": "question_type"
                                                        },
                                                        {
                                                            "type": "boolean",
                                                            "name": "required",
                                                            "default": false
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                {
                                                                    "type": "long",
                                                                    "logicalType": "timestamp-micros"
                                                                }
                                                            ],
                                                            "name": "min_value",
                                                            "default": null
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                {
                                                                    "type": "long",
                                                                    "logicalType": "timestamp-micros"
                                                                }
                                                            ],
                                                            "name": "max_value",
                                                            "default": null
                                                        }
                                                    ]
                                                },
                                                {
                                                    "type": "record",
                                                    "name": "_timepicker_options",
                                                    "namespace": "vela.model.questionnaire",
                                                    "fields": [
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_question_type_timepicker",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "TIMEPICKER"
                                                                ]
                                                            },
                                                            "name": "question_type"
                                                        },
                                                        {
                                                            "type": "boolean",
                                                            "name": "required",
                                                            "default": false
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                {
                                                                    "type": "long",
                                                                    "logicalType": "time-micros"
                                                                }
                                                            ],
                                                            "name": "min_value",
                                                            "default": null
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                {
                                                                    "type": "long",
                                                                    "logicalType": "time-micros"
                                                                }
                                                            ],
                                                            "name": "max_value",
                                                            "default": null
                                                        }
                                                    ]
                                                },
                                                {
                                                    "type": "record",
                                                    "name": "_number_options",
                                                    "namespace": "vela.model.questionnaire",
                                                    "fields": [
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_question_type_number",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "NUMBER"
                                                                ]
                                                            },
                                                            "name": "question_type"
                                                        },
                                                        {
                                                            "type": "boolean",
                                                            "name": "required",
                                                            "default": false
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                "double"
                                                            ],
                                                            "name": "min_value",
                                                            "default": null
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                "double"
                                                            ],
                                                            "name": "max_value",
                                                            "default": null
                                                        },
                                                        {
                                                            "type": "long",
                                                            "name": "precision",
                                                            "default": 0
                                                        },
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_number_widget",
                                                                "namespace": "vela.model.questionnaire",
                                                                "symbols": [
                                                                    "INPUT",
                                                                    "INPUT_WITH_BUTTONS"
                                                                ]
                                                            },
                                                            "name": "widget"
                                                        }
                                                    ]
                                                }
                                            ],
                                            "name": "options"
                                        },
                                        {
                                            "type": {
                                                "type": "array",
                                                "items": {
                                                    "type": "record",
                                                    "name": "flow_rule",
                                                    "namespace": "vela.model.types.questionnaire",
                                                    "fields": [
                                                        {
                                                            "type": "string",
                                                            "name": "rule_code"
                                                        },
                                                        {
                                                            "type": {
                                                                "type": "enum",
                                                                "name": "_flow_rule_type",
                                                                "namespace": "vela.model.types.questionnaire",
                                                                "symbols": [
                                                                    "SKIP",
                                                                    "JUMP"
                                                                ]
                                                            },
                                                            "name": "rule_type"
                                                        },
                                                        {
                                                            "type": [
                                                                "null",
                                                                "long"
                                                            ],
                                                            "name": "target_questionnaire_node_internal_id",
                                                            "default": null
                                                        }
                                                    ]
                                                }
                                            },
                                            "name": "flow_rules"
                                        },
                                        {
                                            "type": {
                                                "type": "long",
                                                "logicalType": "timestamp-micros"
                                            },
                                            "name": "created_at"
                                        },
                                        {
                                            "type": {
                                                "type": "long",
                                                "logicalType": "timestamp-micros"
                                            },
                                            "name": "updated_at"
                                        },
                                        {
                                            "type": "long",
                                            "name": "created_by"
                                        },
                                        {
                                            "type": "long",
                                            "name": "updated_by"
                                        }
                                    ]
                                },
                                {
                                    "type": "record",
                                    "name": "_questionnaire_node",
                                    "namespace": "vela.model.questionnaire",
                                    "fields": [
                                        {
                                            "type": "long",
                                            "name": "questionnaire_node_internal_id"
                                        },
                                        {
                                            "type": {
                                                "type": "enum",
                                                "name": "_questionnaire_node_type_generic",
                                                "namespace": "vela.model.questionnaire",
                                                "symbols": [
                                                    "SECTION",
                                                    "GROUP"
                                                ]
                                            },
                                            "name": "questionnaire_node_type"
                                        },
                                        {
                                            "type": "string",
                                            "name": "label"
                                        },
                                        {
                                            "type": [
                                                "null",
                                                {
                                                    "type": "map",
                                                    "values": "string"
                                                }
                                            ],
                                            "name": "text",
                                            "default": null
                                        },
                                        {
                                            "type": {
                                                "type": "array",
                                                "items": "vela.model.types.questionnaire.flow_rule"
                                            },
                                            "name": "flow_rules"
                                        },
                                        {
                                            "type": {
                                                "type": "long",
                                                "logicalType": "timestamp-micros"
                                            },
                                            "name": "created_at"
                                        },
                                        {
                                            "type": {
                                                "type": "long",
                                                "logicalType": "timestamp-micros"
                                            },
                                            "name": "updated_at"
                                        },
                                        {
                                            "type": "long",
                                            "name": "created_by"
                                        },
                                        {
                                            "type": "long",
                                            "name": "updated_by"
                                        },
                                        {
                                            "type": {
                                                "type": "array",
                                                "items": [
                                                    "vela.model.questionnaire._questionnaire_node",
                                                    "vela.model.questionnaire._question"
                                                ]
                                            },
                                            "name": "children",
                                            "default": []
                                        }
                                    ]
                                }
                            ]
                        },
                        "name": "nodes"
                    },
                    {
                        "type": {
                            "type": "long",
                            "logicalType": "timestamp-micros"
                        },
                        "name": "created_at"
                    },
                    {
                        "type": {
                            "type": "long",
                            "logicalType": "timestamp-micros"
                        },
                        "name": "updated_at"
                    },
                    {
                        "type": "long",
                        "name": "created_by"
                    },
                    {
                        "type": "long",
                        "name": "updated_by"
                    }
                ]
            },
            "name": "body"
        }
    ]
}
""")

message = {
    'header': OrderedDict([
        ('message_uuid', '6e644c49-89d0-4858-aabe-b02ae944e709'),
        ('message_source', 'questionnaire_api'),
        ('producer_ip_addresses', ['127.0.0.1']),
        ('message_created_at', datetime.now()),
        ('partner_internal_id', 39302), ('organization_internal_id', 39302),
        ('business_event_type', 'QUESTIONNAIRE_UPDATED')
    ]),
    'body': OrderedDict([
        ('created_at', datetime.now()),
        ('updated_at', datetime.now()),
        ('created_by', 177033),
        ('updated_by', 177033),
        ('questionnaire_internal_id', 733),
        ('label', 'adawdaw'),
        ('display_label', {'en': 'adawdaw'}),
        ('data_tag', None),
        ('nodes', [
            {
                'created_at': datetime.now(),
                'updated_at': datetime.now(),
                'created_by': 177033,
                'updated_by': 177033,
                'questionnaire_node_internal_id': 734,
                'label': 'how_are_you_adawdaw',
                'text': {'en': 'How are you doing today? Choose an answer.'},
                'flow_rules': [

                ],
                'children': [],
                'questionnaire_node_type': 'QUESTION',
                'data_tag': 'UPMC',
                'options': {
                    'widget': 'RADIO',
                    'choices': [
                        OrderedDict([('value', 'good'), ('text', {'en': 'I had a good day.'})]),
                        OrderedDict([('value', 'ok'), ('text', {'en': 'I had an ok day – ups and downs.'})]),
                        OrderedDict([('value', 'struggled'), ('text', {'en': 'I struggled – today was a hard day.'})])
                    ],
                    'required': True,
                    'columns_number': 1,
                    'question_type': 'CHOICE'
                }
            }
        ])
    ])
}

validate(message, schema)
```

**Comment by rsichny** [comment]
also, this was fixed by removing `children` field from the data, so wondering if that's expected behavior now.

**Comment by scottbelden** [comment]
Thanks for the report. You are right that the change you mentioned is what is causing this to fail.

Looking at your example, the message has a node array can either be a `vela.model.questionnaire._question` record or a `vela.model.questionnaire._questionnaire_node` record. However, based on the actual data you have, I'm not sure which one it is. See below:

```python
('nodes', [
            {
                'created_at': datetime.now(),
                'updated_at': datetime.now(),
                'created_by': 177033,
                'updated_by': 177033,
                'questionnaire_node_internal_id': 734,
                'label': 'how_are_you_adawdaw',
                'text': {'en': 'How are you doing today? Choose an answer.'},
                'flow_rules': [

                ],
                'children': [],  # <---------- This field should only exist on a vela.model.questionnaire._questionnaire_node record
                'questionnaire_node_type': 'QUESTION',
                'data_tag': 'UPMC',  # <---------- This field should only exist on a vela.model.questionnaire._question
                'options': {
                    'widget': 'RADIO',
                    'choices': [
                        OrderedDict([('value', 'good'), ('text', {'en': 'I had a good day.'})]),
                        OrderedDict([('value', 'ok'), ('text', {'en': 'I had an ok day – ups and downs.'})]),
                        OrderedDict([('value', 'struggled'), ('text', {'en': 'I struggled – today was a hard day.'})])
                    ],
                    'required': True,
                    'columns_number': 1,
                    'question_type': 'CHOICE'
                }
            }
        ])
```

The record you have has both a `children` and `data_tag` field defined which doesn't look like it should match any of the potential records in the schema. I think before the most recent change, `fastavro` was ignoring the fact that the `children` field existed and saying that your data matched the `vela.model.questionnaire._question` record type.

**Comment by rsichny** [comment]
@scottbelden seems like in 0.21.5 fastavro was deciding based on just validation instead of fields presence (and new version accounts for both), and so main reason for making decision which schema to use was `questionnaire_node_type` field's value which can not be valid for both schemas.

**Comment by rsichny** [comment]
i suppose the proper fix to this should be the following:
1. validate the record against each schema in the union
2. if the record passed validation against one (and only one schema) - use this schema
3. if the record passed validation against multiple schemas - check the fields.

**Comment by scottbelden** [comment]
Are we considering records with extra fields to be valid?

**Comment by rsichny** [comment]
hmm, just checked, you're not... now i'm now not sure why this worked in previous version :/

anyway, as i said, my issue is resolved, if you feel like this is expected change in behavior - just close this.

**Comment by scottbelden** [comment]
Correct. However, it should be noted that this more strict checking that now takes place will in practice only be checked when unions are involved (as validation currently only occurs when determining union schemas) unless someone specifically calls `validate` like you did. For example, this still works:

```python
from fastavro import writer, reader
from io import BytesIO

schema = { 
    "type": "record",
    "name": "test",
    "fields": [
        {"type": "string", "name": "name"}
    ]   
}

records = [{"name": "abc", "other": "asd"}]

bio = BytesIO()

writer(bio, schema, records)

bio.seek(0)

for record in reader(bio):
    print(record)
```

The output would be `{u'name': u'abc'}`.

Other implementations (for example Ruby) are [implementing a `strict` option](https://github.com/apache/avro/pull/321) that would cause my example to fail.

**Comment by scottbelden** [comment]
The more strict checking only occurs when unions are involved (which explains why it fails in your example [here](https://github.com/fastavro/fastavro/issues/274#issuecomment-422080077) but not in [my example](https://github.com/fastavro/fastavro/issues/274#issuecomment-422071688).

The strict checking should probably be across the board (so that my example would also fail), but I fear that doing that could be a pretty large breaking change and so I would want to be more cautious there. The change to make unions more strict was of course also a breaking change, but the previous behavior would cause some pretty awful results as seen here: https://github.com/fastavro/fastavro/issues/272#issue-360336581

**Comment by rsichny** [comment]
yeah, i get it, so maybe consider impletenting this instead: https://github.com/fastavro/fastavro/issues/274#issuecomment-421995222 and add another option for strict mode?

PS. sorry for close/reopen - misclick

**Comment by scottbelden** [comment]
> if the record passed validation against one (and only one schema) - use this schema

Is a schema with extra fields considered "passing" validation?

**Comment by scottbelden** [comment]
Actually, I think I have a simple solution. I looked back at the previous problem and realized that it's failing for a slightly different reason than I thought. So I think I can fix that case and have your original test case also work.

**Comment by scottbelden** [comment]
I pushed https://github.com/fastavro/fastavro/pull/275

It essentially reverts https://github.com/fastavro/fastavro/pull/273 so that field names are not checked but fixes the actual problem from https://github.com/fastavro/fastavro/issues/272 which is that if a field wasn't defined then it would try to get the default value and if that didn't exist it would return `None` which is a problem as `None` is a valid data type. Now an object representing `no_value` will be used and will cause the validation to fail in cases where it was silently passing.

When running with https://github.com/fastavro/fastavro/pull/275 I was able to get the existing tests to pass as well as the test case you mentioned in this issue.
